WebSockets Next: introduce OpenConnections

mkouba opened the pull request:
See https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/websockets-next.3A.20more.20questions

CC @turing85

mkouba committed:
WebSockets Next: introduce OpenConnections

- also add WebSocket#endpointId() and WebSocketConnection#endpointId()

mkouba commented on extensions/websockets-next/server/runtime/src/main/java/io/quarkus/websockets/next/OpenConnections.java:
```diff
@@ -0,0 +1,36 @@
+ * @see WebSocketConnection
+ */
+@Experimental("This API is experimental and may change in the future")
+public interface OpenConnections {
```
Alternatively, we could provide a bean of type `Set<WebSocketConnection>`. This would have to be a stateless delegating proxy though, because the underlying set of open connections is dynamic. And injecting an immutable snapshot is impractical for beans with wider scope, e.g. `@ApplicationScoped`.

The other downside is the impossibility to define the contract in the javadoc.

geoand replied:
I think `OpenConnections` makes perfect sense

mkouba commented on extensions/websockets-next/server/runtime/src/main/java/io/quarkus/websockets/next/WebSocketConnection.java:
```diff
@@ -29,6 +29,12 @@ public interface WebSocketConnection extends Sender, BlockingSender {
+     *
+     * @return the fully qualified class name of the {@link WebSocket} endpoint
+     */
+    String endpoint();
```
This is more like an _endpoint id_, I can change the javadoc. I'd like to avoid using `Class<?>` here.

mkouba replied:
Well, the method name does not contain the `class` at all :wink:.

But yes, we could improve this. The main goal of this "property" is to allow users to filter connections of a specific endpoint. `Class<?>` seems to be an obvious solution but I'd like to try to avoid loading the class here, hence the string.

mkouba replied:
Hm, that's actually a nice idea. I'd call it endpoint id. And it could be defaulted to FQCN.

turing85 replied:
Exactly what I had in mind :wink:

mkouba replied:
Done.

turing85 replied:
Nice! Thanks!

gastaldi commented on extensions/websockets-next/server/runtime/src/main/java/io/quarkus/websockets/next/OpenConnections.java:
```diff
@@ -0,0 +1,36 @@
+     *
+     * @return an immutable collection of all open connections
+     */
+    Collection<WebSocketConnection> get();
```
Why this method when you already have the `stream()` below that does the same practically?

turing85 replied:
With what java version do we build? If we build with `>= 17`, we can call [`stream().toList()` (`docs.oracle.com`)](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/stream/Stream.html#toList()).

gastaldi commented on extensions/websockets-next/server/runtime/src/main/java/io/quarkus/websockets/next/OpenConnections.java:
```diff
@@ -0,0 +1,36 @@
+     * @return the open connection or {@code null} if no open connection with the given id exists
+     * @see WebSocketConnection#id()
+     */
+    WebSocketConnection get(String id);
```
Is this `id` the `endpointId` or the `WebSocketConnection#id`?

mkouba replied:
It's `WebSocketConnection#id()`, see javadoc. Also it would not make sense to obtain a single connection for an endpoint. But I agree that it would not hurt to rename `id` to `connectionId`. Will do it ;-)

gastaldi reviewed (approved):
LGTM. I'd prefer to avoid the utility methods reusing the Stream since that can be done in the user's code, but that can be revisited in a later PR.

gastaldi commented on extensions/websockets-next/server/runtime/src/main/java/io/quarkus/websockets/next/OpenConnections.java:
```diff
@@ -0,0 +1,44 @@
+     * @return the open connection or {@code null} if no open connection with the given id exists
+     * @see WebSocketConnection#id()
+     */
+    default WebSocketConnection get(String connectionId) {
```
```suggestion
    default Optional<WebSocketConnection> get(String connectionId) {
```

gastaldi replied:
Maybe renaming to `findByConnectionId` could be a good idea too, to indicate the purpose

gastaldi commented on extensions/websockets-next/server/runtime/src/main/java/io/quarkus/websockets/next/OpenConnections.java:
```diff
@@ -0,0 +1,44 @@
+     * @see WebSocketConnection#id()
+     */
+    default WebSocketConnection get(String connectionId) {
+        return stream().filter(c -> c.id().equals(connectionId)).findFirst().orElse(null);
```
```suggestion
        return stream().filter(c -> c.id().equals(connectionId)).findFirst();
```

mkouba replied:
But this would require the return type to be `Optional<WebSocketConnection>`, or?

turing85 replied:
Yes, see https://github.com/quarkusio/quarkus/pull/39858#discussion_r1550549920

mkouba replied:
Hm, we don't use `Optional` anywhere in the API yet, so I'm not quite sure :shrug:.

gastaldi commented on extensions/websockets-next/server/runtime/src/main/java/io/quarkus/websockets/next/OpenConnections.java:
```diff
@@ -0,0 +1,44 @@
+     *
+     * @return an immutable collection of all open connections
+     */
+    default Collection<WebSocketConnection> get() {
```
```suggestion
    default Collection<WebSocketConnection> listAll() {
```

turing85 commented on extensions/websockets-next/server/runtime/src/main/java/io/quarkus/websockets/next/OpenConnections.java:
```diff
@@ -0,0 +1,44 @@
+     * @return an immutable collection of all open connections
+     */
+    default Collection<WebSocketConnection> get() {
+        return stream().collect(Collectors.toUnmodifiableList());
```
```suggestion
        return stream().toList();
```

(If we are on java `>= 17`)

mkouba commented:
> LGTM. I'd prefer to avoid the utility methods reusing the Stream since that can be done in the user's code, but that can be revisited in a later PR.

Well, I think that one of our goals is to help users write less code and these are typical operations a user might need, so why not add these tiny methods? :shrug:

Anyway, I've renamed the methods, made the `OpenConnections` extend `Iterable<WebSocketConnection>` and added `findByEndpointId()`... so hopefully it's good to go now :wink:.

gastaldi commented on extensions/websockets-next/server/runtime/src/main/java/io/quarkus/websockets/next/runtime/ConnectionManager.java:
```diff
@@ -1,27 +1,40 @@
 
+    @Override
+    public Iterator<WebSocketConnection> iterator() {
+        return stream().iterator();
```
This can be moved as a default method in the `OpenConnections` interface

mkouba replied:
It could but for some reason I don't think it's appropriate. Let's keep it here for now...
